igvc_navigation/src/path_follower/SmoothControl.cpp

```diff
@@ -35,81 +31,57 @@ void SmoothControl::getTrajectory(igvc_msgs::velocity_pair& vel, nav_msgs::Path&
     pose.pose.position.x = end_pos[0];
     pose.pose.position.y = end_pos[1];
     trajectory.poses.push_back(pose);
-    // std::cout << std::endl;
     cur_pos = end_pos;
   }
-  // std::cout << "\n end" << std::endl;
 }
 
-void SmoothControl::getAction(Eigen::Vector3d& result, Eigen::Vector3d cur_pos, Eigen::Vector3d target)
+void SmoothControl::getAction(Eigen::Vector3d& result, Eigen::Vector3d cur_pos,
+        Eigen::Vector3d target, Eigen::Vector2d egocentric_heading)
 {
-  double delta, theta;
-  Eigen::Vector3d slope = target - cur_pos;
-  // std::cout << "atan2 = " << atan2(slope[1], slope[0]) << std::endl;
-  if (slope[0] == 0)
-  {
-    delta = slope[1] > 0 ? -M_PI / 2 : M_PI / 2;
-    delta += cur_pos[2];
-  }
-  else
-  {
-    delta = -atan2(slope[1], slope[0]) + cur_pos[2];
-  }
-  while (delta > M_PI)
-  {
-    delta -= 2 * M_PI;
-  }
-  while (delta < -M_PI)
-  {
-    delta += 2 * M_PI;
-  }
+  /**
+  Computes the radius of curvature to obtain a new angular velocity value.
+  */
 
-  slope = cur_pos - target;
-  if (slope[0] == 0)
-  {
-    theta = slope[1] > 0 ? M_PI / 2 : -M_PI / 2;
-    theta -= target[2];
-  }
-  else
-  {
-    theta = atan2(slope[1], slope[0]) - target[2];
-  }
-  theta += M_PI;
-  theta = -theta;
-  while (theta > M_PI)
-  {
-    theta -= 2 * M_PI;
-  }
-  while (theta < -M_PI)
-  {
-    theta += 2 * M_PI;
-  }
+  // get egocentric polar coordinates for delta and theta
+  double delta = egocentric_heading[0], theta = egocentric_heading[1];
 
-  // std::cout << "slope = "<< slope[0] << ", " << slope[1] << std::endl;
-  // std::cout << "theta = " << theta << " delta = " << delta << std::endl;
+  // adjust both angles to lie between -PI and PI
+  fitToPolar(delta);
+  fitToPolar(theta);
 
+  // calculate the radius of curvature, K
   double d = sqrt(pow(target[0] - cur_pos[0], 2) + pow(target[1] - cur_pos[1], 2));
-
-  double K = k2 * (delta - atan(k1 * theta));
+  double K = k2 * (delta - atan(-k1 * theta));
```

Comment:
Did you mean to negate the theta? This should flip the quadrants we are operating in, but I'm not sure why that is necessary.

Reply:
I'm fairly certain this is the correct implementation. I think this has to do with the fact that delta and theta are egocentric (turn left is + , right is -), hence the negation.